cumulusci/core/dependencies/dependencies.py

```diff
@@ -0,0 +1,910 @@
+            )
+
+        context.logger.warn(
+            f"No 2GP package version located for release {release_id} on {repo.clone_url}."
+        )
+        return (None, None)
+
+
+RESOLVER_CLASSES = {
+    DependencyResolutionStrategy.STRATEGY_STATIC_TAG_REFERENCE: GitHubTagResolver,
+    DependencyResolutionStrategy.STRATEGY_2GP_EXACT_BRANCH: GitHubReleaseBranchExactMatch2GPResolver,
+    DependencyResolutionStrategy.STRATEGY_2GP_RELEASE_BRANCH: GitHubReleaseBranch2GPResolver,
+    DependencyResolutionStrategy.STRATEGY_2GP_PREVIOUS_RELEASE_BRANCH: GitHubPreviousReleaseBranch2GPResolver,
+    DependencyResolutionStrategy.STRATEGY_BETA_RELEASE_TAG: GitHubBetaReleaseTagResolver,
+    DependencyResolutionStrategy.STRATEGY_RELEASE_TAG: GitHubReleaseTagResolver,
+    DependencyResolutionStrategy.STRATEGY_UNMANAGED_HEAD: GitHubUnmanagedHeadResolver,
+}
+
+
+## External API
+
+
+def get_resolver(
+    strategy: DependencyResolutionStrategy, dependency: DynamicDependency
+) -> Optional[Resolver]:
+    # This will be fleshed out when further types of DynamicDependency are added.
+
+    return RESOLVER_CLASSES[strategy]()
+
+
+def get_resolver_stack(
+    context: BaseProjectConfig, name: str
+) -> List[DependencyResolutionStrategy]:
+    resolutions = context.project__dependency_resolutions
+    stacks = context.project__dependency_resolutions__resolution_strategies
+
+    if name in resolutions and name != "resolution_strategies":
+        name = resolutions[name]
+
+    if stacks and name in stacks:
+        return [DependencyResolutionStrategy(n) for n in stacks[name]]
+
+    raise CumulusCIException(f"Resolver stack {name} was not found.")
+
+
+def get_static_dependencies(
+    dependencies: List[Dependency],
+    strategies: List[DependencyResolutionStrategy],
+    context: BaseProjectConfig,
+    ignore_deps: List[dict] = None,
+):
+    """Resolves the project__dependencies section of cumulusci.yml
+    to convert dynamic GitHub dependencies into static dependencies
+    by inspecting the referenced repositories.
+
+    Keyword arguments:
+    :param dependencies: a list of dependencies to resolve
+    :param ignore_deps: if provided, ignore the specified dependencies wherever found.
+    """
+
+    while any(not d.is_flattened or not d.is_resolved for d in dependencies):
+        for d in dependencies:
+            if not d.is_resolved:
+                d.resolve(context, strategies)
+
+        def unique(it: Iterable):
+            seen = set()
+
+            for each in it:
+                if each not in seen:
+                    seen.add(each)
+                    yield each
+
+        dependencies = list(
+            unique(
+                itertools.chain(
+                    *list(
+                        d.flatten(context)
+                        for d in dependencies
+                        if not _should_ignore_dependency(d, ignore_deps or [])
+                    )
+                ),
+            )
+        )
```

Comment:
I think we could probably come up with a better name than `flatten`. It's not very evocative of being the function that finds transitive dependencies.